internal/manage/devices/download.go

```diff
@@ -0,0 +1,337 @@
+package devices
+
+import (
+	"context"
+	"encoding/base64"
+	"encoding/csv"
+	"fmt"
+	"iter"
+	"maps"
+	"net/http"
+	"net/url"
+	"slices"
+	"strings"
+
+	"google.golang.org/grpc/codes"
+	"google.golang.org/grpc/status"
+	"google.golang.org/protobuf/proto"
+	"google.golang.org/protobuf/reflect/protopath"
+	"google.golang.org/protobuf/reflect/protorange"
+	"google.golang.org/protobuf/reflect/protoreflect"
+
+	"github.com/smart-core-os/sc-api/go/traits"
+	"github.com/smart-core-os/sc-golang/pkg/resource"
+	"github.com/smart-core-os/sc-golang/pkg/trait"
+	"github.com/vanti-dev/sc-bos/pkg/gen"
+)
+
+//go:generate protomod protoc -- -I . -I ../../../proto --go_out=paths=source_relative:. download.proto
+
+func (s *Server) GetDownloadDevicesUrl(_ context.Context, request *gen.GetDownloadDevicesUrlRequest) (*gen.DownloadDevicesUrl, error) {
+	// validate
+	switch request.MediaType {
+	case "":
+		request.MediaType = "text/csv"
+	case "text/csv": // supported
+	default:
+		return nil, status.Errorf(codes.InvalidArgument, "unsupported media type %q", request.MediaType)
+	}
+
+	tokenData := &DownloadToken{
+		Request: request,
+	}
+	tokenStr, err := encodeDownloadToken(tokenData)
+	if err != nil {
+		return nil, err
+	}
+	u := s.downloadUrlBase
+	if err := s.encodeDownloadUrlToken(&u, tokenStr); err != nil {
+		return nil, err
+	}
+	return &gen.DownloadDevicesUrl{
+		Url: u.String(),
+	}, nil
+}
+
+// DownloadDevicesHTTPHandler responds to HTTP request urls returned by GetDownloadDevicesUrl.
+func (s *Server) DownloadDevicesHTTPHandler(w http.ResponseWriter, r *http.Request) {
```

Comment:
This is a big method that combines business logic with HTTP handling.
I would prefer to see a separate `func GenerateDevicesCSV(w io.Writer, devices []*traits.Metadata) error` (or something like that) with unit tests.

Reply:
Have a look at that, I think it's better